File: frontend/src/types/review.ts
export interface Issue {
  severity: 'critical' | 'high' | 'medium' | 'low'
  file: string
  line: number | null
  category: 'bug' | 'security' | 'performance' | 'style'
  description: string
  suggestion: string
  current_code: string
  proposed_code: string
  confidence: number
  priority: 'must_fix' | 'should_fix' | 'nice_to_fix'
}

export interface FileReview {
  file: string
  summary: string
  issues: Issue[]
  suggestions: string[]
}

export interface RewrittenFile {
  filename: string
  language: string
  content: string
  issues_fixed: number
}

export interface UsageMetrics {
  total_time_s: number
  llm_time_s: number
  input_tokens: number
  output_tokens: number
  rate_limit_remaining: number | null
}

export interface ReviewResult {
  owner: string
  repo: string
  pull_number: number
  pr_title: string
  pr_description: string
  pr_merged: boolean
  files_changed: number
  additions: number
  deletions: number
  risk_level: 'high' | 'medium' | 'low'
  summary: string
  file_reviews: FileReview[]
  issues: Issue[]
  suggestions: string[]
  scores: Record<string, number>
  share_token: string
  github_review_id: number | null
  rewritten_files: RewrittenFile[]
  usage: UsageMetrics | null
}

export interface ReviewProgress {
  phase: 'fetching' | 'chunking' | 'reviewing' | 'reviewing_security' | 'reviewing_normal' | 'summarizing'
  current: number
  total: number
  file?: string
  language?: string
  message?: string
}

export interface TrendEntry {
  id: number
  owner: string
  repo: string
  pr_title: string
  pull_number: number
  risk_level: string
  issue_count: number
  suggestion_count: number
  files_changed: number
  additions: number
  deletions: number
  scores: Record<string, number>
  created_at: string
}

export interface FileInfo {
  filename: string
  language: string
  patch: string
}

export interface ModelInfo {
  provider: string
  model: string
}

export type ReviewPhase = 'idle' | 'loading' | 'progress' | 'streaming' | 'done' | 'error'

export interface ProviderInfo {
  name: string
  display_name: string
  is_builtin: boolean
  is_enabled: boolean
  default_model: string
  models: string[]
  needs_config: boolean
  base_url?: string
}

export interface CustomProviderInput {
  name: string
  display_name: string
  base_url: string
  api_key: string
  default_model: string
  timeout?: number
}

File: backend/models/review.py
from pydantic import BaseModel


class FileChange(BaseModel):
    filename: str
    status: str  # added | modified | removed | renamed
    patch: str = ""
    additions: int = 0
    deletions: int = 0
    language: str = ""
    context_hint: str = ""  # 分片上下文：如 "class UserService — 方法 create_user()"


class PRInfo(BaseModel):
    owner: str
    repo: str
    pull_number: int
    title: str = ""
    description: str = ""
    files: list[FileChange] = []
    additions: int = 0
    deletions: int = 0
    head_sha: str = ""  # PR head commit SHA，用于获取完整文件内容


class Issue(BaseModel):
    severity: str  # critical | high | medium | low
    file: str
    line: int | None = None
    category: str  # bug | security | performance | style
    description: str
    suggestion: str = ""
    # 1.2: 建议具体化 — LLM 输出 current/proposed 代码对比
    current_code: str = ""   # 当前有问题的代码片段
    proposed_code: str = ""  # 建议修改后的代码片段
    # 1.3: 置信度 + 优先级
    confidence: int = 0   # LLM 自评置信度 0-100
    priority: str = "should_fix"  # must_fix | should_fix | nice_to_fix


class FileReview(BaseModel):
    file: str
    summary: str = ""
    issues: list[Issue] = []
    suggestions: list[str] = []


class ReviewResult(BaseModel):
    owner: str
    repo: str
    pull_number: int
    pr_title: str = ""
    files_changed: int = 0
    additions: int = 0
    deletions: int = 0
    risk_level: str = "low"
    summary: str = ""
    file_reviews: list[FileReview] = []
    issues: list[Issue] = []
    suggestions: list[str] = []
    # 3.1: 量化评分
    scores: dict = {}  # { overall, security, bug, performance, style } 0-100
    # 4.1: 分享
    share_token: str = ""
    # 4.3: GitHub PR Review 回写
    github_review_id: int | None = None
